### PR TITLE
try to build a callback to retrieve stored data live

### LINKED ISSUE #1079: Doc/question: Adaptive scans with area detectors
Reading through the docs for single process [adaptive scans](http://nsls-ii.github.io/bluesky/tutorial.html#adaptive-logic-in-a-plan) everything looks great! However, does the same logic work for area detectors which save files? I'm guessing that the results that the detector responds with will be the filename and/or other information that is passed to the databroker for it to properly register the data. If this is the case what changes are needed to that scan to work with file saving detectors?

### PR DESCRIPTION
**Pull request by CJ-Wright**
Add a callback which keeps track of resource and datum documents, and has a retrieve method which can be used to turn datum ids into actual data.

closes #1079

### CODE REVIEW

File: bluesky/callbacks/core.py [hunk 1]
```diff
@@ -351,3 +352,46 @@ def start(self, doc):
     def _print(self, out_str):
         self._rows.append(out_str)
         self._out(out_str)
+
+
+class Retrieve(CallbackBase):
+    def __init__(self, root_map=None, handler_reg=None):
```
Comment (danielballan) [inline bluesky/callbacks/core.py]: Why is the handler registry optional?

File: bluesky/callbacks/core.py [hunk 2]
```diff
@@ -351,3 +352,46 @@ def start(self, doc):
...
+    def start(self, doc):
+        self.resources = {}
+        self.handlers = {}
+        self.datums = {}
+
+    def resource(self, resource):
+        self.resources[resource['uid']] = resource
+        handler = self.handler_reg[resource['spec']]
+
+        key = (str(resource['uid']), handler.__name__)
```
Comment (danielballan) [inline bluesky/callbacks/core.py]: I think I see what you're getting at here, but I'm not sure this is the best way to achieve that. What else could we key on?

File: bluesky/callbacks/core.py [hunk 3]
```diff
@@ -351,3 +352,46 @@ def start(self, doc):
...
+        if handler_reg is None:
+            handler_reg = {}
+        self.resources = None
+        self.handlers = None
+        self.datums = None
+        self.root_map = root_map
+        self.handler_reg = handler_reg
+
+    def start(self, doc):
+        self.resources = {}
```
Comment (danielballan) [inline bluesky/callbacks/core.py]: These are never cleared. Should they be?

File: bluesky/callbacks/core.py (hunk 2 above)
Comment (CJ-Wright) [reply bluesky/callbacks/core.py]: I pulled this out of the databroker code.

File: bluesky/callbacks/core.py (hunk 3 above)
Comment (CJ-Wright) [reply bluesky/callbacks/core.py]: What do you mean they are never cleared? I guess I only wanted to deal with data from the run in question, so I cleared things out.

**Comment by CJ-Wright** [comment]
This most likely should also implement a `fill_event` method.

File: bluesky/callbacks/core.py (hunk 1 above)
Comment (CJ-Wright) [reply bluesky/callbacks/core.py]: Again copied out of the databroker code. I can make it required though.

File: bluesky/callbacks/core.py (hunk 2 above)
Comment (CJ-Wright) [reply bluesky/callbacks/core.py]: https://github.com/NSLS-II/databroker/blob/master/databroker/assets/base_registry.py#L249

**Comment by CJ-Wright** [comment]
I think this is ready for another round of review.

**Comment by CJ-Wright** [comment]
It might be good to have `LiveImage` subclass from this. That way a) we remove the databroker dep and b) we can scrub the images, since we can cache the events and an index and then submit them as the "new" event to the update method, at which point lookup/retrieval will happen.

**Comment by klauer** [comment]
Seems like a useful callback handler to have. 👍 (Can't comment on the implementation details just yet, though - still getting back into the whole NSLS-II stack)

**Comment by danielballan** [comment]
Power-cycling...

File: bluesky/callbacks/core.py (hunk 3 above)
Comment (danielballan) [reply bluesky/callbacks/core.py]: If it is meant to be a long-lived instance, the `resources` can grow large.

**Comment by CJ-Wright** [comment]
Should this return documents, so that it is more like `databroker.documents(fill=True)`

File: bluesky/callbacks/core.py (hunk 2 above)
Comment (danielballan) [reply bluesky/callbacks/core.py]: OK, then let's improve it there as well. I'm concerned that this isn't unique enough. What do you think about:

```py
(str(resource['uid']), handler.__module__, handler.__qualname__)
```

The `__module__` is the one I'm worried about omitting, but we might as well upgrade to `__qualname__` while we're here. Context: https://www.python.org/dev/peps/pep-3155/

Comment (CJ-Wright) [reply bluesky/callbacks/core.py]: I like `__qualname__`

File: bluesky/callbacks/core.py (hunk 3 above)
Comment (CJ-Wright) [reply bluesky/callbacks/core.py]: Hmm. Is making the dict anew with every start not enough to clear the resources? Or should this be cleared more frequently? If so, how often?

Comment (danielballan) [reply bluesky/callbacks/core.py]: You're right. Now I have no idea what I was thinking with that original comment.

**Comment by danielballan** [comment]
This point has come up a couple times, so we must be talking past each other. What do you mean when you say "return"? It's a callback, so `return` is a dead end.

File: bluesky/callbacks/core.py (hunk 1 above)
Comment (danielballan) [reply bluesky/callbacks/core.py]: Fair. But yes let's make it required.

**Comment by CJ-Wright** [comment]
The return could be convenient for the `LiveImage` rewrite, as we could have `event = super().event(doc)` where the new event would be filled. This would also be helpful for data processing pipelines of any kind since filling events is an important part of the data processing.

File: bluesky/callbacks/core.py [hunk 4]
```diff
@@ -351,3 +356,74 @@ def start(self, doc):
...
+        self.handlers = {}
+        self.datums = {}
+
+    def resource(self, resource):
+        self.resources[resource['uid']] = resource
+        handler = self.handler_reg[resource['spec']]
+
+        key = (str(resource['uid']), handler.__name__)
```
Comment (CJ-Wright) [inline bluesky/callbacks/core.py]:
```suggestion
        key = (str(resource['uid']), handler.__module__, handler.__qualname__)
```

**Comment by CJ-Wright** [comment]
This code should really poke the resources and datums at a lower level.

**Comment by danielballan** [comment]
I think we should aim for something like [this](https://gist.github.com/danielballan/cd16936debeb03a674ac1671fe476a0b#file-sketch_analysisstore-py-L33-L72), added to the `event_model` package -- used by both databroker 1.0 and any applicable callbacks in bluesky -- for implementing filling. Maybe that's what you have in mind as well.

**Comment by CJ-Wright** [comment]
Sorry this was the wrong PR to close. I think having both the document and datum access is useful.

**Comment by danielballan** [comment]
Not sure what that means, can you explain?

**Comment by CJ-Wright** [comment]
Having this method would be helpful: https://github.com/NSLS-II/bluesky/pull/1099/files#diff-692eb92ad1a097aaa8ea5c83893ad2e6R396

**Comment by CJ-Wright** [comment]
I think this exists now

**Pull request closed without merging**